### PR TITLE
[POST] /auth/signup

### LINKED ISSUE #13: [POST]  auth/signup
Create route for sign up form

### PR DESCRIPTION
**Pull request by Ibtisam-Hemmo**

### CODE REVIEW

**Review by Nader-SH — DISMISSED** [review]

File: server/validation/signupSchema.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  owner_name:string,
+    owner_id: number,
+    name:string,
+    license_number:number,
+    location:string,
+    phone:number,
+    email:string,
+    password:string,
+    confirmPassword:string
+}
```
Comment (Raghad-Mezied) [inline server/validation/signupSchema.ts]: Move to interfaces

File: server/controllers/authentication/signUp.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    if (emailExisted) {
+      throw new CustomError(400, 'Try again, This email is already existed');
+    }
+
+    const hashed = await hash(req.body.password, 10);
+    const pharamcyData = await signup(req.body, hashed);
+
+    const token = await generateToken({
+      owner_id: pharamcyData.owner_id,
+      role: 'pharmacy',
```
Comment (Raghad-Mezied) [inline server/controllers/authentication/signUp.ts]: We need the image and the pharmacy_id not the owner_id

File: server/controllers/authentication/signUp.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    });
+
+    return res.cookie('token', token).json(pharamcyData);
```
Comment (Raghad-Mezied) [inline server/controllers/authentication/signUp.ts]: Send it with the status code 201, and send the json with the pattern `json({
  data,
  msg
})`

File: server/controllers/authentication/signUp.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  } catch (err) {
+    next(err);
+  }
```
Comment (Raghad-Mezied) [inline server/controllers/authentication/signUp.ts]: This error could be the validation error, so we need to send it with the customError you did

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

File: server/controllers/authentication/signUp.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    const token = await generateToken({
+      id: pharamcyData.id,
+      role: 'pharmacy',
+      owner_img: pharamcyData.owner_img
+    });
+
+    return res.cookie('token', token).json({data: pharamcyData, msg: 'You have signed up successfully'});
+  } catch (err) {
+      if(err.name === 'ValidationError') {
+        next(new CustomError (401, 'you have a validation error when signing up'))
```
Comment (Raghad-Mezied) [inline server/controllers/authentication/signUp.ts]: It's a validation error, the user enters something wrong so send an appropriate message and status code.

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Raghad-Mezied — DISMISSED** [review]

**Review by Nader-SH — DISMISSED** [review]

**Review by Raghad-Mezied — DISMISSED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Review by Nader-SH — APPROVED** [review]

**Review by AhmedAbuSharar — APPROVED** [review]

**Pull request merged**